packages/vite-plugin-rsc/src/plugin.ts

```diff
@@ -0,0 +1,212 @@
+      if (node.declaration?.type === "VariableDeclaration") {
+        node.declaration?.declarations?.forEach((declaration) => {
+          const name = declaration?.id?.name;
+          if (name) {
+            exports.push(name);
+          }
+        });
+        return;
+      }
+
+      if (node.declaration?.type === "FunctionDeclaration") {
+        const name = node.declaration?.id?.name;
+        if (name) {
+          exports.push(name);
+        }
+        return;
+      }
+
+      if (node.specifiers?.length) {
+        node.specifiers.forEach((specifier) => {
+          const name = specifier?.exported?.name;
+          if (name) {
+            exports.push(name);
+          }
+        });
+      }
+    }
+  });
+  return exports;
+};
+
+export default function plugin({
+  serverDist = "dist/server",
+  clientDist = "dist/static",
+}: {
+  serverDist?: string;
+  clientDist?: string;
+}): Plugin {
+  const clientPragma = "use client";
+  // const serverPragma = "use server";
+  let externals = new Set<string>();
+  let config: ResolvedConfig;
+  let isSsr: boolean;
+  let isBuild: boolean;
+  let manifest: Manifest = {};
+
+  const bundleMap = new Map();
+  const clientModuleId = "virtual:client-bundle-map";
+  const resolvedClientModuleId = "\0" + clientModuleId;
+  const serverModuleId = "virtual:server-bundle-map";
+  const resolvedServerModuleId = "\0" + serverModuleId;
+
+  const clientBundleMapFilename = "client-bundle-map.json";
+  const serverBundleMapFilename = "server-bundle-map.json";
+
+  return {
+    name: "vite-plugin-extract-server-components",
+
+    config(config) {
+      config.build ||= {};
+      if (config.build.ssr) {
+        const manifestPath = path.join(
+          config.root || "",
+          clientDist,
+          "manifest.json"
+        );
+        if (existsSync(manifestPath)) {
+          manifest = JSON.parse(readFileSync(manifestPath, "utf-8"));
+        }
+        config.build.rollupOptions ||= {};
+        config.build.rollupOptions.external = (id) => externals.has(id);
+      }
+    },
+
+    configResolved(resolvedConfig) {
+      config = resolvedConfig;
+      isSsr = !!config.build.ssr;
+      isBuild = config.command === "build";
+    },
+    resolveId(id, source) {
+      if (id === clientModuleId) {
+        return resolvedClientModuleId;
+      }
+      if (id === serverModuleId) {
+        return resolvedServerModuleId;
+      }
+    },
+    load(id) {
+      if (id === resolvedClientModuleId) {
+        return `export default ${JSON.stringify(
+          // Yes the client bundle map is in the server dist because
+          // it's the SSR build that generates the client bundle map
+          path.join(config.root || "", serverDist, clientBundleMapFilename)
+        )}`;
+      }
+      if (id === resolvedServerModuleId) {
+        return `export default ${JSON.stringify(
+          path.join(config.root || "", clientDist, serverBundleMapFilename)
+        )}`;
+      }
+    },
```

Comment:
These are virtual modules that just provide the path to the module map JSON files. This lets the prerender script or server load the JSON file without needing to know the path. Instead it can import the module and get the filename.